### ISSUE TITLE
Circular reference misdetection

### ISSUE THREAD

**Issue by falsandtru** [microsoft/TypeScript#32950]
**TypeScript Version:**  3.4.0-dev.20190817

**Search Terms:**

**Code**

```ts
export function f<T extends HTMLOListElement | DocumentFragment>(target: T): T {
  if (target instanceof HTMLOListElement) {
    for (const el of target.querySelectorAll('a')) {
      if (el.closest('ol') !== target) continue;
    }
  }
  return target;
}
```

**Expected behavior:**
pass

**Actual behavior:**
```
const el: any
'el' implicitly has type 'any' because it does not have a type annotation and is referenced directly or indirectly in its own initializer.
```

**Playground Link:** http://www.typescriptlang.org/play/index.html#code/KYDwDg9gTgLgBAMwK4DsDGMCWEWIDwAqcoMwKAJgM5wASBAsgDIDyjmlMAogDbAC2ZeAB84AEQhokAlDABiUAIYBzaTAB8AChgKoS4DABccAgEojRAN4AoOHEwI4WnXviYUHBemAQHdJq3YuXlUTOGtbWwRoRzQcDmJuOB84bV19ADoARyRgKABPAGVgXgxoAEFubg0AcgVqk1DwiNt7R2L0tG4ISmAOGohuergAQgBeUZTnfVDYmTccgG4bCIBfZbXbKH0kKFxUlyWVoA

**Related Issues:**

**Comment by RyanCavanaugh** [comment]
🤔🤔🤔 Usually there's some sneaky dependency here, but I can't find one. We might be incorrectly thinking that the back edge of the `for` loop somehow influences `el`, but it doesn't?

Simplifiable a bit
```ts
export function f(target: HTMLOListElement | DocumentFragment) {
  if (target instanceof HTMLOListElement) {
    for (const el of target.querySelectorAll('a')) {
      if (el === target) {
      }
    }
  }
}
```

**Comment by falsandtru** [comment]
This bug is found in writing a polyfill of `:scope` keyword for *Edge* browser.

```ts
// expect
if (target instanceof HTMLOListElement) {
  for (const el of target.querySelectorAll(':scope > li > a')) {
  }
}
// polyfill
if (target instanceof HTMLOListElement) {
  for (const el of target.querySelectorAll('a')) {
    if (el.closest('ol') !== target) continue;
  }
}
```

**Comment by ahejlsberg** [comment]
I think what's happening here is that in order to determine the type of `el` we have to determine the type of `target.querySelectorAll('a')`, which requires us determine the control flow type of `target`, which could be narrowed by the expression `el.closest('ol') !== target`, which requires us to determine the type of `el`, which triggers a circularity.

**Comment by thorn0** [comment]
Another instance ([playground](http://www.typescriptlang.org/play/#code/CYUwxgNghgTiAEAzArgOzAFwJYHtVIAoBKALnlWQFsAjEGeAH3jVES1RGAG4AoMPAM4Z4AbSgAaeNXgBeeFAC6s0QEYADJMTEFvIA); reproduces in TS 3.6.2 too):
```ts
declare function f(): number | undefined;
const [a, b = a] = [10, f()];
// 'a' implicitly has type 'any' because it does not have a type annotation and 
// is referenced directly or indirectly in its own initializer.
```

**Comment by falsandtru** [comment]
Not similar to the others. Should be filed as a separated issue.

**Comment by danvk** [comment]
This was fixed in d26afd7273c8aabfb92157df55eadbc907c6fb65 which was part of #32695, which was released in TS 3.7 back in 2019.
